### DISCUSSION TITLE
New Language Support

### DISCUSSION THREAD

**Discussion by Frogmore42** [arendst/Tasmota#10128] [Q&A] [answered]
@ziotprojects asked how to support Thai language

https://github.com/arendst/Tasmota/discussions/10124#discussioncomment-202932

**Comment by Frogmore42** [comment]
This is a good document on the ISO standard for language and country codes: https://docs.oracle.com/cd/E13214_01/wli/docs92/xref/xqisocodes.html

Normally language codes are lower case and country/region codes are uppercase.  Tasmota has quite a few here:
https://github.com/arendst/Tasmota/tree/development/tasmota/language

But, I don't see th_TH in the list.
to use uppercase for the language code and I don't see TH in the list.

If you can do the translation from one of the other languages, you can follow the example, like this one:
https://github.com/arendst/Tasmota/blob/development/tasmota/language/uk_UA.h

**Reply by ziotprojects** [reply]
Thanks, I am working on it.

**Comment by arendst** [comment, marked as answer]
As was documented above the following steps are needed to provide a Pull Request (PR) for a new language like Thai.

- Visit website https://www.science.co.il/language/Locale-codes.php to find the correct ``LCID string`` representing your country and language. In case of Thai in Thailand it would become ``th_TH``
- Copy any language file from the language folder you feel comfortable with to translate and give it the name ``th_TH.h``.

Start editing the file with following changes:
- Change ``#define LANGUAGE_LCID`` to the correct ``LCID decimal``. In case of Thailand it would be 1054
- Change ``#define D_HTML_LANGUAGE`` to the correct ``Language code``. In case of Thai it would be "th"
- Translate ``#define D_DAY3LIST`` to 7 days of the week, starting with sunday and having exactly three characters a day.
- Translate ``#define D_MONTH3LIST`` to 12 months of the year, starting with january and having exactly three characters a month.
- Translate all other defines keeping them as short as possible as every character takes up code space.

Provide the PR and we will implement your language by executing the following steps:

Add compile information to below named files:
- Find the localization area within file ``my_user_config.h`` and add a line like ``//#define MY_LANGUAGE            th_TH           // Thai in Thailand``
- Edit file ``platformio_tasmota_env.ini`` by inserting the following lines:
```
[env:tasmota-TH]
build_flags             = ${common.build_flags} -DMY_LANGUAGE=th_TH
```
- Edit file ``platformio_tasmota_env32.ini`` by inserting the following lines:
```
[env:tasmota32-TH]
build_flags             = ${common32.build_flags} -DMY_LANGUAGE=th_TH
```
- Edit files ``platform.ini`` and ``platform_tasmota32.ini`` by adding below ``default_envs =`` a line like 
```
;                tasmota-TH
```

Add language specific information to workflow files in folder ``.github/workflows`` by scrolling through 
``CI_github.yml``, ``CI_github_ESP32.yml`` and ``Tasmota_build.yml`` and copy any language section. Then edit this section with your specific language information. Compare different languages to find what minimal changes are needed.

This will provide compiled binaries which in turn need to be registered by the OTA download website.

**Reply by ziotprojects** [reply]
Thanks, I am working on it.
